File: 205. Isomorphic Strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        dics = {}
        dict = {}
        for i in range(len(s)):
            if s[i] in dics and t[i] in dict:
                if dics[s[i]] != t[i]:
                    return False
                if dict[t[i]] != s[i]:
                    return False
            elif s[i] not in dics and t[i] not in dict:
                dics[s[i]] = t[i]
                dict[t[i]] = s[i]
            elif t[i] not in dict:
                if dics[s[i]] != t[i]:
                    return False
                dict[t[i]] = s[i]
            else:
                if dict[t[i]] != s[i]:
                    return False
                dics[s[i]] = t[i]
        return True

File: 205. Isomorphic Strings.java
class Solution {
    public boolean isIsomorphic(String s, String t) {
        Map<Character, Character> stot = new HashMap<>();
        Map<Character, Character> ttos = new HashMap<>();
        for (int i=0; i<s.length(); i++) {
            char cs = s.charAt(i);
            char ct = t.charAt(i);
            if (stot.containsKey(cs)) {
                if (stot.get(cs) != ct) {
                    return false;
                }
            } else {
                stot.put(cs, ct);
            }
            if (ttos.containsKey(ct)) {
                if (ttos.get(ct) != cs) {
                    return false;
                }
            } else {
                ttos.put(ct, cs);
            }
        }
        return true;
    }
}

class Solution {
    public boolean isIsomorphic(String s, String t) {
        int[] stot = new int[256];
        int[] ttos = new int[256];
        for (int i=0; i<s.length(); i++) {
            char cs = s.charAt(i);
            char ct = t.charAt(i);
            if (stot[cs] != ttos[ct]) {
                return false;
            }
            stot[cs] = i+1;
            ttos[ct] = i+1;
        }
        return true;
    }
}